### ISSUE TITLE
Module does not compile anymore with perl 5.21.7 (prototype changes?)

### ISSUE THREAD

**Issue by eserte** [oalders/lwp-consolelogger#3]
See http://www.cpantesters.org/cpan/report/7e98a350-8a7a-11e4-a381-d2b4dfbfc7aa for a sample fail report. Probably relevant perldelta entry is https://metacpan.org/pod/release/CORION/perl-5.21.7/pod/perldelta.pod#and-prototypes-accepts-only-subs
If you think that this is not a valid change in perl, then please speak up at perl5-porters.

**Comment by garu** [comment]
@oalders tl;dr: there are two issues in play, both easily fixable:

1) always pass variables to p().
2) make sure you'll always get the 'return' behaviour of p(), even if it's overwritten by the user (in a .dataprinter file for instance).

A simple patch is replacing that line with:

```perl
my $json = decode_json( $content );
$content = p($json, return_value => 'dump');
```

another solution (should give you the same result):

```perl
my $json = decode_json( $content );
undef $content; # <-- otherwise p() will append to the current string
p($json, output => \$content);
```
Now the long comment :)

This is quite interesting! It exposes a known bug in Data::Printer where settings aren't localized. For instance, if elsewhere in the code (say, by some app using your module) someone uses Data::Printer with another set of configs, they might override your "global" settings with their own. This was reported by frew and rjbs and I'm working on implementing an internal object to fix it in the near future.

More importantly in terms of usage, it goes to show that there are people using Data::Printer not just to dump data on their debug sessions but also to format data for users as well. This is important because in order to fix another bug I'm going to need to change Data::Printer's default return value from 'dump' to 'pass', and that will break code using it the way you were (before my patch above). I wish I didn't have to change the return value's default behaviour, but I can't think of a better solution. Now, because this is not the first time I heard of people using p() to return data, I'm experimenting with a new function, np(), which will always return the formatted string and will be ideal exactly for those use cases.

Could you please take a look at https://github.com/garu/Data-Printer/issues/16 ? It explains the issue and solutions, including one that's already on CPAN as a trial release. This means that, if that solution is accepted by the community, your proper fix (instead of the workaround fixes I pasted above) will be simply to replace calls to ```p()``` with calls to ```np()```.

Thanks!

**Comment by oalders** [comment]
@garu Wow -- thanks very much for the detailed explanation.  :+1: 

np() seems like the way to go for me.  Easy fix. I'll wait for the full release and then patch my code.

**Comment by eserte** [comment]
FYI: it fails also with 0.000014: http://matrix.cpantesters.org/?dist=LWP-ConsoleLogger%200.000014

**Comment by oalders** [comment]
@eserte thanks! I'm just waiting for a new Data::Printer to be released so that I can fix this.

**Comment by oalders** [comment]
So, waiting on a new release was silly, since there was already a proposed fix from @garu.  I've implemented that in 5841ce77, so this should be fixed now.  I was getting a lot of failure reports via smokers from @eserte.  I noticed that they had the latest TRIAL release, so that actually encouraged me to test out the TRIAL and just fix it.  Found that to be quite helpful.
